perf-event/src/events.rs

```diff
@@ -318,3 +335,172 @@ pub enum CacheResult {
+/// ```
+///
+/// # Usage Notes
+/// - Some systems do not support creating read-only or write-only breakpoints.
+///   If you are getting `EINVAL` errors while trying to build such a counter
+///   using a read-write breakpoint might work instead.
+///
+/// - It is not valid to have a breakpoint that matches on both read/write and
+///   execute accesses. Trying to do this will result in an error.
+///
+/// - The valid values of len are quite limited. The [`perf_event_open`][man]
+///   manpage indicates that the only valid values for `bp_len` are 1, 2, 4,
+///   and 8.
+///
+/// [man]: http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Breakpoint {
+    /// The type of breakpoint to set.
+    pub ty: BreakpointType,
+
+    /// The address of the breakpoint. For execution breakpoints this is the
+    /// address of the instruction of interest; for read and write breakpoints
+    /// it is the address of the memory location of interest.
+    pub addr: u64,
+
+    /// The length of the breakpoint being measured.
+    ///
+    /// There are a limited number of valid values for this field. Basically,
+    /// the options are 1, 2, 4, and 8. Setting this field to anything else
+    /// will cause counter creation to fail with an error.
+    pub len: u64,
+}
+
+impl Breakpoint {
+    /// Create a breakpoint configuration to count the number of times that
+    /// the instruction at the provided address was executed.
+    pub const fn execute(addr: u64) -> Self {
+        Self {
+            ty: BreakpointType::EXECUTE,
+            addr,
+            // According to the perf_event_open man page, execute breakpoints
+            // should set len to sizeof(long).
+            len: std::mem::size_of::<libc::c_long>() as _,
+        }
+    }
+
+    /// Create a breakpoint configuration to count the number of times that
+    /// we read from the provided memory location.
+    ///
+    /// See the struct field docs for valid values of `len`.
+    pub const fn read(addr: u64, len: u64) -> Self {
+        Self {
+            ty: BreakpointType::READ,
+            addr,
+            len,
+        }
+    }
+
+    /// Create a breakpoint configuration to count the number of times that
+    /// we write to the provided memory location.
+    ///
+    /// See the struct field docs for valid values of `len`.
+    pub const fn write(addr: u64, len: u64) -> Self {
+        Self {
+            ty: BreakpointType::WRITE,
+            addr,
+            len,
+        }
+    }
+
+    /// Create a breakpoint configuration to count the number of times that
+    /// we either read from or write to the provided memory location.
+    ///
+    /// See the struct field docs for valid values of `len`.
+    pub const fn read_write(addr: u64, len: u64) -> Self {
+        Self {
+            ty: BreakpointType::READ_WRITE,
+            addr,
+            len,
+        }
+    }
+}
```

Comment:
All this could go away - we'd just let people construct `Breakpoint` variants.

Reply:
I've left these in for now since I think that, as a user of the API, they make it easier to use. If you really want them gone then I'll remove them.